src/vs/editor/contrib/stickyScroll/browser/stickyScrollProvider.ts

```diff
@@ -0,0 +1,177 @@
+import { ILanguageFeaturesService } from 'vs/editor/common/services/languageFeatures';
+import { OutlineModel, OutlineElement, OutlineGroup } from 'vs/editor/contrib/documentSymbols/browser/outlineModel';
+import { CancellationToken, CancellationTokenSource, } from 'vs/base/common/cancellation';
+import { EditorOption, RenderLineNumbersType } from 'vs/editor/common/config/editorOptions';
+import { RunOnceScheduler } from 'vs/base/common/async';
+import { Range } from 'vs/editor/common/core/range';
+import { Emitter } from 'vs/base/common/event';
+
+export class StickyRange {
+	constructor(
+		public readonly startLineNumber: number,
+		public readonly endLineNumber: number
+	) { }
+}
+
+export class StickyLineCandidate {
+	constructor(
+		public readonly startLineNumber: number,
+		public readonly endLineNumber: number,
+		public readonly nestingDepth: number,
+	) { }
+}
+
+export class StickyLineCandidateProvider extends Disposable {
+	private readonly onStickyScrollChangeEmitter = this._register(new Emitter<void>());
+	public readonly onStickyScrollChange = this.onStickyScrollChangeEmitter.event;
+
+	static readonly ID = 'store.contrib.stickyScrollController';
+	private readonly editor: ICodeEditor;
+	private readonly languageFeaturesService: ILanguageFeaturesService;
+	private readonly updateSoon: RunOnceScheduler;
+
+	private cts: CancellationTokenSource | undefined;
+	private outlineModel: StickyOutlineElement | undefined;
+	private readonly sessionStore: DisposableStore = new DisposableStore();
+	private modelVersionId: number = 0;
+
+	constructor(
+		editor: ICodeEditor,
+		@ILanguageFeaturesService _languageFeaturesService: ILanguageFeaturesService,
+	) {
+		super();
+		this.editor = editor;
+		this.languageFeaturesService = _languageFeaturesService;
+		this.updateSoon = this._register(new RunOnceScheduler(() => this.update(true), 50));
+		this._register(this.editor.onDidChangeConfiguration(e => {
+			if (e.hasChanged(EditorOption.experimental)) {
+				this.readConfiguration();
+			}
+		}));
+		this.readConfiguration();
+	}
+
+	private readConfiguration() {
+		const options = this.editor.getOption(EditorOption.experimental);
+		if (options.stickyScroll.enabled === false) {
+			this.sessionStore.clear();
+			return;
+		} else {
+			this.sessionStore.add(this.editor.onDidChangeModel(() => this.update(true)));
+			this.sessionStore.add(this.editor.onDidScrollChange(() => this.update(false)));
+			this.sessionStore.add(this.editor.onDidChangeHiddenAreas(() => this.update(true)));
+			this.sessionStore.add(this.editor.onDidChangeModelContent(() => this.updateSoon.schedule()));
+			this.sessionStore.add(this.languageFeaturesService.documentSymbolProvider.onDidChange(() => this.update(true)));
+			const lineNumberOption = this.editor.getOption(EditorOption.lineNumbers);
+			if (lineNumberOption.renderType === RenderLineNumbersType.Relative) {
+				this.sessionStore.add(this.editor.onDidChangeCursorPosition(() => this.update(false)));
```

Comment:
I think this is only about rendering sticky lines, thus it should not be in the StickyScrollProvider which just provides data.